cognite/client/data_classes/hosted_extractors/sources.py

```diff
@@ -0,0 +1,636 @@
+            return self._set(value)
+
+    class _PortUpdate(CognitePrimitiveUpdate):
+        def set(self, value: int | None) -> _MQTTUpdate:
+            return self._set(value)
+
+    class _AuthenticationUpdate(CognitePrimitiveUpdate):
+        def set(self, value: MQTTAuthentication | None) -> _MQTTUpdate:
+            return self._set(value.dump() if value else None)
+
+    class _UseTlsUpdate(CognitePrimitiveUpdate):
+        def set(self, value: bool) -> _MQTTUpdate:
+            return self._set(value)
+
+    class _CACertificateUpdate(CognitePrimitiveUpdate):
+        def set(self, value: CACertificate | None) -> _MQTTUpdate:
+            return self._set(value.dump() if value else None)
+
+    class _AuthCertificateUpdate(CognitePrimitiveUpdate):
+        def set(self, value: AuthCertificate | None) -> _MQTTUpdate:
+            return self._set(value.dump() if value else None)
+
+    @property
+    def host(self) -> _HostUpdate:
+        return _MQTTUpdate._HostUpdate(self, "host")
+
+    @property
+    def port(self) -> _PortUpdate:
+        return _MQTTUpdate._PortUpdate(self, "port")
+
+    @property
+    def authentication(self) -> _AuthenticationUpdate:
+        return _MQTTUpdate._AuthenticationUpdate(self, "authentication")
+
+    @property
+    def useTls(self) -> _UseTlsUpdate:
+        return _MQTTUpdate._UseTlsUpdate(self, "useTls")
+
+    @property
+    def ca_certificate(self) -> _CACertificateUpdate:
+        return _MQTTUpdate._CACertificateUpdate(self, "caCertificate")
+
+    @property
+    def auth_certificate(self) -> _AuthCertificateUpdate:
+        return _MQTTUpdate._AuthCertificateUpdate(self, "authCertificate")
+
+    @classmethod
+    def _get_update_properties(cls, item: CogniteResource | None = None) -> list[PropertySpec]:
+        return [
+            PropertySpec("host", is_nullable=False),
+            PropertySpec("port", is_nullable=True),
+            PropertySpec("authentication", is_nullable=True, is_container=True),
+            PropertySpec("useTls", is_nullable=False),
+            PropertySpec("ca_certificate", is_nullable=True, is_container=True),
+            PropertySpec("auth_certificate", is_nullable=True, is_container=True),
+        ]
+
+
+class MQTT3SourceWrite(_MQTTSourceWrite):
+    _type = "mqtt3"
+
+
+class MQTT5SourceWrite(_MQTTSourceWrite):
+    _type = "mqtt5"
+
+
+class MQTT3Source(_MQTTSource):
+    _type = "mqtt3"
+
+
+class MQTT5Source(_MQTTSource):
+    _type = "mqtt5"
+
+
+class MQTT3SourceUpdate(_MQTTUpdate):
+    _type = "mqtt3"
+
+
+class MQTT5SourceUpdate(_MQTTUpdate):
+    _type = "mqtt5"
+
+
+class SourceWriteList(CogniteResourceList[SourceWrite], ExternalIDTransformerMixin):
+    _RESOURCE = SourceWrite
+
+
+class SourceList(WriteableCogniteResourceList[SourceWrite, Source], ExternalIDTransformerMixin):
+    _RESOURCE = Source
+
+    def as_write(
+        self,
+    ) -> SourceWriteList:
+        raise TypeError(f"{type(self).__name__} cannot be converted to write")
```

Comment:
```suggestion
    def as_write(
        self,
    ) -> NoReturn:
        raise TypeError(f"{type(self).__name__} cannot be converted to write because?")
```